### ISSUE TITLE
No operator "[]" matches these operands for large arrays.

### ISSUE THREAD

**Issue by scarrazza** [cupy/cupy#5367]
When allocating a large arrays (with size > 2**27), the [] operator raises an error if the assigned value is different from zero.

### Conditions

<details>
OS                           : Linux-4.15.0-143-generic-x86_64-with-glibc2.10
CuPy Version                 : 8.6.0
NumPy Version                : 1.19.2
SciPy Version                : 1.6.2
Cython Build Version         : 0.29.22
CUDA Root                    : /usr/local/cuda
CUDA Build Version           : 11000
CUDA Driver Version          : 11000
CUDA Runtime Version         : 11000
cuBLAS Version               : 11200
cuFFT Version                : 10101
cuRAND Version               : 10101
cuSOLVER Version             : (10, 2, 0)
cuSPARSE Version             : 11101
NVRTC Version                : (11, 0)
Thrust Version               : 100909
CUB Build Version            : 100909
cuDNN Build Version          : 8101
cuDNN Version                : 8005
NCCL Build Version           : 2804
NCCL Runtime Version         : 2804
cuTENSOR Version             : None
Device 0 Name                : Tesla V100-DGXS-32GB
Device 0 Compute Capability  : 70
Device 1 Name                : Tesla V100-DGXS-32GB
Device 1 Compute Capability  : 70
Device 2 Name                : Tesla V100-DGXS-32GB
Device 2 Compute Capability  : 70
Device 3 Name                : Tesla V100-DGXS-32GB
Device 3 Compute Capability  : 70
</details>

### Code to reproduce
```python
import cupy as cp
x = cp.zeros(2**28, dtype=cp.complex128)
x[0] = 1
```
### Error messages, stack traces, or logs

<details>
In [4]: x = cp.zeros(2**28, dtype=cp.complex128)

In [5]: x[0]=1
---------------------------------------------------------------------------
NVRTCError                                Traceback (most recent call last)
~/.local/lib/python3.9/site-packages/cupy/cuda/compiler.py in compile(self, options, log_stream)
    548                     nvrtc.addAddNameExpression(self.ptr, ker)
--> 549             nvrtc.compileProgram(self.ptr, options)
    550             mapping = None

cupy_backends/cuda/libs/nvrtc.pyx in cupy_backends.cuda.libs.nvrtc.compileProgram()

cupy_backends/cuda/libs/nvrtc.pyx in cupy_backends.cuda.libs.nvrtc.compileProgram()

cupy_backends/cuda/libs/nvrtc.pyx in cupy_backends.cuda.libs.nvrtc.check_status()

NVRTCError: NVRTC_ERROR_COMPILATION (6)

During handling of the above exception, another exception occurred:

CompileException                          Traceback (most recent call last)
<ipython-input-5-7215f83d98c8> in <module>
----> 1 x[0]=1

cupy/core/core.pyx in cupy.core.core.ndarray.__setitem__()

cupy/core/_routines_indexing.pyx in cupy.core._routines_indexing._ndarray_setitem()

cupy/core/_routines_indexing.pyx in cupy.core._routines_indexing._scatter_op()

cupy/core/core.pyx in cupy.core.core.ndarray.fill()

cupy/core/_kernel.pyx in cupy.core._kernel.ElementwiseKernel.__call__()

cupy/core/_kernel.pyx in cupy.core._kernel.ElementwiseKernel._get_elementwise_kernel()

cupy/_util.pyx in cupy._util.memoize.decorator.ret()

cupy/core/_kernel.pyx in cupy.core._kernel._get_elementwise_kernel()

cupy/core/_kernel.pyx in cupy.core._kernel._get_simple_elementwise_kernel()

cupy/core/_kernel.pyx in cupy.core._kernel._get_simple_elementwise_kernel()

cupy/core/core.pyx in cupy.core.core.compile_with_cache()

~/.local/lib/python3.9/site-packages/cupy/cuda/compiler.py in compile_with_cache(source, options, arch, cache_dir, extra_source, backend, enable_cooperative_groups, name_expressions, log_stream)
    368             name_expressions, log_stream, cache_in_memory)
    369     else:
--> 370         return _compile_with_cache_cuda(
    371             source, options, arch, cache_dir, extra_source, backend,
    372             enable_cooperative_groups, name_expressions, log_stream,

~/.local/lib/python3.9/site-packages/cupy/cuda/compiler.py in _compile_with_cache_cuda(source, options, arch, cache_dir, extra_source, backend, enable_cooperative_groups, name_expressions, log_stream, cache_in_memory)
    433     if backend == 'nvrtc':
    434         cu_name = '' if cache_in_memory else name + '.cu'
--> 435         ptx, mapping = compile_using_nvrtc(
    436             source, options, arch, cu_name, name_expressions,
    437             log_stream, cache_in_memory)

~/.local/lib/python3.9/site-packages/cupy/cuda/compiler.py in compile_using_nvrtc(source, options, arch, filename, name_expressions, log_stream, cache_in_memory)
    209                 cu_file.write(source)
    210 
--> 211             return _compile(source, options, cu_path,
    212                             name_expressions, log_stream)
    213     else:

~/.local/lib/python3.9/site-packages/cupy/cuda/compiler.py in _compile(source, options, cu_path, name_expressions, log_stream)
    193                              name_expressions=name_expressions)
    194         try:
--> 195             ptx, mapping = prog.compile(options, log_stream)
    196         except CompileException as e:
    197             dump = _get_bool_env_variable(

~/.local/lib/python3.9/site-packages/cupy/cuda/compiler.py in compile(self, options, log_stream)
    558         except nvrtc.NVRTCError:
    559             log = nvrtc.getProgramLog(self.ptr)
--> 560             raise CompileException(log, self.src, self.name, options,
    561                                    'nvrtc' if not runtime.is_hip else 'hiprtc')
    562 

CompileException: /home/xxx/.local/lib/python3.9/site-packages/cupy/core/include/cupy/carray.cuh(185): error: the size of an array must be greater than zero
          detected during instantiation of class "CArray<T, _ndim, _c_contiguous, _use_32bit_indexing> [with T=thrust::complex<double>, _ndim=0, _c_contiguous=true, _use_32bit_indexing=false]" 
/tmp/tmpz50ag5oy/3a4362b2326b13391ecae5492c6008c1_2.cubin.cu(8): here

/home/xxx/.local/lib/python3.9/site-packages/cupy/core/include/cupy/carray.cuh(186): error: the size of an array must be greater than zero
          detected during instantiation of class "CArray<T, _ndim, _c_contiguous, _use_32bit_indexing> [with T=thrust::complex<double>, _ndim=0, _c_contiguous=true, _use_32bit_indexing=false]" 
/tmp/tmpz50ag5oy/3a4362b2326b13391ecae5492c6008c1_2.cubin.cu(8): here

/home/xxx/.local/lib/python3.9/site-packages/cupy/core/include/cupy/carray.cuh(202): error: the size of an array must be greater than zero
          detected during instantiation of class "CArray<T, _ndim, _c_contiguous, _use_32bit_indexing> [with T=thrust::complex<double>, _ndim=0, _c_contiguous=true, _use_32bit_indexing=false]" 
/tmp/tmpz50ag5oy/3a4362b2326b13391ecae5492c6008c1_2.cubin.cu(8): here

/home/xxx/.local/lib/python3.9/site-packages/cupy/core/include/cupy/carray.cuh(207): error: the size of an array must be greater than zero
          detected during instantiation of class "CArray<T, _ndim, _c_contiguous, _use_32bit_indexing> [with T=thrust::complex<double>, _ndim=0, _c_contiguous=true, _use_32bit_indexing=false]" 
/tmp/tmpz50ag5oy/3a4362b2326b13391ecae5492c6008c1_2.cubin.cu(8): here

/tmp/tmpz50ag5oy/3a4362b2326b13391ecae5492c6008c1_2.cubin.cu(12): error: no operator "[]" matches these operands
            operand types are: CArray<thrust::complex<double>, 0, true, false> [ const ptrdiff_t * ]

5 errors detected in the compilation of "/tmp/tmpz50ag5oy/3a4362b2326b13391ecae5492c6008c1_2.cubin.cu".

</details>

**Comment by emcastillo** [comment]
Hi, 
Thank you for your report!

This issue was already filled in #5328
and it's fixed in the current master branch as of #5330
We will be releasing a version this Thursday with this fix included!

**Comment by scarrazza** [comment]
@emcastillo great thanks!
